### PR TITLE
Application portal - Save form data

### PR DESCRIPTION
**Pull request by w-gao**
Problem
=======
Load/Save application form [Jira](https://cruzhacks-dev.atlassian.net/browse/C2D-137)

Solution
========
What I/we did to solve this problem
* Implement `SaveData()`
* Also switched from localStorage to sessionStorage 

Change Summary:
---------------
* Added interfaces for saved props in props.ts
* Updated applicationForm to support saving on next click
* Updated Debug Button to use sessionStorage
* Updated storage.ts + tests to use sessionStorage

Dev-Ops
=======
If you added an ENV variable, please check off that you've updated the following  
- [ ] Key & Sample value to `.env` & `sample.env` 
- [ ] GCP Secret Manager (Both development and production)
- [ ] Github Secrets (Added a development and production variable)
- [ ] Github Workflows `.github/workflows/dev.yml` & `.github/workflows/prod.yml`
- [ ] webpack-config.js

Images/Important Notes (optional):
-----------------------
* Insert any notes/issues, dependencies added or removed, or images

### CODE REVIEW

File: src/components/ApplicationContext/ApplicationContext.tsx
Comment (krishnakpandian) [inline src/components/ApplicationContext/ApplicationContext.tsx]: This seems to not save the file locally (which is good since this is probably a security risk and we don't want to store that much data in the browser) but do we know have code that explicitly handles this or is it just working?

**Review by krishnakpandian — COMMENTED** [review]
Tested this and it seems like it works! I did have a question on the functionality of storing data.

File: src/components/ApplicationContext/ApplicationContext.tsx
Comment (krishnakpandian) [inline src/components/ApplicationContext/ApplicationContext.tsx]: Can we also put a timer on this so it doesn't stay indefinitely in our user's localstorage.

**Review by krishnakpandian — CHANGES_REQUESTED** [review]
Did some more testing to make sure different accounts can't access data and I think we should add a timeLimit to the data persistence. (Maybe like a month)

File: src/components/ApplicationContext/ApplicationContext.tsx
Comment (w-gao) [reply src/components/ApplicationContext/ApplicationContext.tsx]: This was definitely an oversight on my part. I did not consider the case with the resume, but it looks like the `File` object serializes to `{}` and doesn't fail when we do `JSON.stringify()` (at least on Chrome), but we should definitely explicitly remove that field.

I'm also noticing that this stores the error messages too, which is OK I guess unless we want to strip those out as well.

File: src/components/ApplicationContext/ApplicationContext.tsx
Comment (w-gao) [reply src/components/ApplicationContext/ApplicationContext.tsx]: Yep! A month would be `60 * 60 * 24 * 30` seconds right? Just wanted to make sure...

Also, the timeout is just for our logic, it'll still stay in the localStorage if the user doesn't visit our site after a month. So we should probably clear this after the user submits, and/or after the application deadline.

Comment (krishnakpandian) [reply src/components/ApplicationContext/ApplicationContext.tsx]: I think that's a good call and for our onSubmit we can handle that

File: src/components/ApplicationContext/ApplicationContext.tsx
Comment (krishnakpandian) [reply src/components/ApplicationContext/ApplicationContext.tsx]: We can probably just strip the error messages and file. If you want, we can modify the current interface to extend of the fields that you need for storing and retrieving.

File: src/components/ApplicationContext/ApplicationContext.tsx
Comment (krishnakpandian) [inline src/components/ApplicationContext/ApplicationContext.tsx]: I think one thing we should do is run the validation functions before saving. For example if a user inputs binary data, that will be stored in localstorage. If you run the validation functions with the onSubmit argument set to False it should validate only fields that must always be true (i.e valid alphanumeric data)

Comment (krishnakpandian) [reply src/components/ApplicationContext/ApplicationContext.tsx]: I was able to save+load emojis/non standard chars

**Review by krishnakpandian — CHANGES_REQUESTED** [review]
Tested again and everything still seems functional. I think one last change.

**Comment by krishnakpandian** [comment]
As a side note, we may want to store appStatus with a user in localstorage so other users don't steal the appStatus.

File: src/components/ApplicationContext/ApplicationContext.tsx
Comment (w-gao) [reply src/components/ApplicationContext/ApplicationContext.tsx]: That makes sense! Will do that after today's meeting.

Should we add a flash message above the form to notify the user if we can't save? It'll probably be useful for our submit function too, if the validation fails for the previous pages. Something like this for example:

(example from the Internet)

Comment (krishnakpandian) [reply src/components/ApplicationContext/ApplicationContext.tsx]: Something like that could work! Could we also hold of some changes on this for now. I've been thinking about future proofing security for 2023 onwards and I wanted to get your thoughts on switching to session storage. If we do switch to session which we'll talk about at today's meeting, we can probably just run the save function onClick of the next/prev button.

Main reason I was thinking of switching is a scenario like this.

Localstorage: User Saves App-> Never submits app -> Data is never cleared -> user never revisits cruzhacks.com for 1 year -> auth0 is off -> stored personal data can never be cleared (unless user manually clears browsers cache) -> Potential XSS vulnerability in 2023 onward or even in current code base (data is compromised) or this could be a problem to future teams using caching (i.e similar naming conventions)

Comment (w-gao) [reply src/components/ApplicationContext/ApplicationContext.tsx]: This is definitely a valid concern. I also don't like storing the application in local storage.

Although, if we don't want to use local storage, perhaps we shouldn't store it at all? If we switch to session storage, I assume we would remove the "save" option so we don't confuse our user. But how often would a user navigate away from the "/portal" route (our ApplicationContext) and come back within the session?

I guess another option is to just ask for confirmation before the user navigates away from the portal view, similar to how Google Forms does it, and not bother with saving anything that is not permanent.

We should talk more about this during today's meeting.

File: src/components/ApplicationContext/ApplicationContext.tsx
Comment (krishnakpandian) [inline src/components/ApplicationContext/ApplicationContext.tsx]: It might be good to maybe save the farthest page visited so we can use that in progress bar on page 0 for a started application. We should probably save this if design thinks it's good and in a separate PR.

**Review by krishnakpandian — APPROVED** [review]
LGTM! Thanks for putting in the effort for this story, I know this one had a bit more internal design indecisiveness!

File: src/components/ApplicationContext/ApplicationContext.tsx
Comment (w-gao) [reply src/components/ApplicationContext/ApplicationContext.tsx]: Oh yeah, that would be cool. There are still a few other things that I want to address before we merge this (unless we need this by Sunday?).

Comment (krishnakpandian) [reply src/components/ApplicationContext/ApplicationContext.tsx]: I would like this by Sunday, but if there are things that need to be addressed we should prioritize that.

**Comment by w-gao** [comment]
A few notes:

* This will still save unvalidated inputs (e.g.: emojis) when user goes to the previous page
  * should we just not save in this case?
* Save farthest page visited https://github.com/CruzHacks/cruzhacks-2022-website/pull/99#discussion_r747787149
  * rethink how we save 
* I also wanted to add a prompt when there are unsaved changes, but we can't differentiate between user changes and our changes easily (unless we introduce more state variables)

Snippet that may be useful:
```typescript
  useEffect(() => {
    const confirmLeave = (ev: BeforeUnloadEvent) => {
      ev.preventDefault()
      // eslint-disable-next-line no-param-reassign
      ev.returnValue = "Are you sure?"
    }

    window.addEventListener("beforeunload", confirmLeave)
    return () => {
      window.removeEventListener("beforeunload", confirmLeave)
    }
  }, [])
```

**Comment by krishnakpandian** [comment]
@w-gao 
1. Save unvalidated inputs we can probably run the save function [here](https://github.com/CruzHacks/cruzhacks-2022-website/blob/848168cd18b2aedd23e8ba700e73074ee7cae5be/src/views/Portal/components/ApplicationForm/index.view.tsx#L47) instead on next so it does validation before saving. As far as validating on prev, I'm not sure what's the best approach here. We could just not save (which might be annoying in the case a user does refresh unsaved data) or we could validation telling them they have on previous but that could run into a weird issue where a user wants to go back a page first before updating the current page. I think having saving just as a backup in case a user needs to refresh their tab should be the goal for us since we aren't supporting cross tab support. I'm leaning more towards option 1 in this situation but we let's chat on this some more because I'm not sure what the best choice would be in this situation. 
2. This can probably be handled really quickly in its own chore
3. I like this idea a lot. I think this can also be handled in a story next sprint. 

For now, I think if we just fix saving unvalidated inputs, that should be enough for testing on the staging on branch.

File: src/Props/portal/application.ts
Comment (w-gao) [inline src/Props/portal/application.ts:4]: Breaking change

File: src/components/ApplicationContext/ApplicationContext.tsx
Comment (w-gao) [inline src/components/ApplicationContext/ApplicationContext.tsx:163]: I changed this to save current page only, so we won't save unvalidated inputs in the other pages.

File: src/views/Portal/components/ApplicationStatus/index.view.tsx
Comment (w-gao) [inline src/views/Portal/components/ApplicationStatus/index.view.tsx:31]: I'll implement this in a separate PR along with the progress bar update once we get design's approval.

**Review by w-gao — COMMENTED** [review]
I made some more changes that require a re-review.

**Comment by w-gao** [comment]
> I think having saving just as a backup in case a user needs to refresh their tab should be the goal for us since we aren't supporting cross tab support. I'm leaning more towards option 1 in this situation but we let's chat on this some more because I'm not sure what the best choice would be in this situation.

@krishnakpandian I agree! Option 1 would be sufficient for saving I think, especially when we add a confirmation popup when users have unsaved data.

**Comment by krishnakpandian** [comment]
> > I think having saving just as a backup in case a user needs to refresh their tab should be the goal for us since we aren't supporting cross tab support. I'm leaning more towards option 1 in this situation but we let's chat on this some more because I'm not sure what the best choice would be in this situation.
> 
> @krishnakpandian I agree! Option 1 would be sufficient for saving I think, especially when we add a confirmation popup when users have unsaved data.

Awesome, if we're on the same page, I think this PR has all the requirements we need for now. We can add in the confirmation popup + update farthest page visited for home page next sprint or even this sprint if there's time. Correct me I'm wrong on this!

**Review by krishnakpandian — APPROVED** [review]
I tested through it and I think it everything seems good

**Comment by w-gao** [comment]
@krishnakpandian Yep, I think this can be merged. I can take those tickets too, but will probably get it done next sprint.

**Pull request merged**
